### PR TITLE
fix: prevent session key collision on disk (#4057)

### LINKED ISSUE #4057: Bug: distinct session keys can collide on disk after filename sanitization
## Summary

Distinct logical session keys can map to the same session file because storage identity uses lossy filename sanitization.

Example collision:

```text
telegram:a_b
telegram:a:b
```

Both can become:

```text
telegram_a_b
```

## Root cause

`SessionManager.safe_key()` / `safe_filename()` replace unsafe characters such as `:` with `_`. This is not reversible and not collision-resistant.

## Failure scenario

1. Two different channels/chats/users produce distinct session keys.
2. Both keys sanitize to the same filename.
3. The two sessions share or overwrite the same persisted history.
4. Users can see contaminated context or lose history.

## Expected behavior

Distinct session keys must always map to distinct storage files.

## Suggested fix

Use a collision-resistant or reversible encoding for storage identity:

- `base64url(raw_session_key)`, or
- `sha256(raw_session_key)` plus metadata sidecar, or
- percent/URL-style reversible encoding.

## Regression test

Assert these session keys map to distinct filenames:

```python
[
    "telegram:a_b",
    "telegram:a:b",
    "telegram/a:b",
    "telegram_a:b",
]
```

### PR DESCRIPTION
**Pull request by axelray-dev**
Fix session key collision on disk (Fixes #4057) without breaking WebUI filename generation.

**Problem:** `SessionManager.safe_key()` replaced `:` with `_`, causing collisions between distinct keys like `telegram:a_b` and `telegram:a:b` (both became `telegram_a_b`). Two sessions would share or overwrite the same persisted history.

**Fix:** Split storage identity from public filename identity:

- `safe_key()` keeps the old lossy encoding (`safe_filename(key.replace(':', '_'))`) -- used by WebUI transcript/thread/session-list paths (transcript.py, thread_disk.py, session_list_index.py).
- New private `_storage_key()` uses base64url (no padding) for collision-resistant session filenames on disk.
- `_get_session_path()` now uses `_storage_key()`, so save/load/delete use collision-resistant paths.
- `_load()` checks new path first, then falls back to the old lossy path in sessions_dir, then the legacy ~/.nanobot/sessions/ path -- existing files are migrated to the new path on first access.
- `delete_session()` cleans up all three possible locations.
- `list_sessions()` and WebUI session list scanning use `_decode_storage_key()` to decode base64url storage stems before falling back to the old lossy guess -- corrupt-file repair works correctly for new filenames.

**Backward compatibility:** Existing session files at the old lossy path are read and migrated to the new collision-resistant path on first access via `_load()`. No data loss. WebUI transcript/thread file names are unchanged.

**Tests:** Added regression tests for distinct filenames, save/load/migration behavior, safe_key collision vs. storage-key collision resistance, and save/load for colliding keys. Updated existing path-matching tests to use `_storage_key()`.

### CODE REVIEW

File: nanobot/session/manager.py [hunk 1]
```diff
@@ -806,15 +838,16 @@ def list_sessions(self) -> list[dict[str, Any]]:
         sessions = []
 
         for path in self.sessions_dir.glob("*.jsonl"):
-            fallback_key = path.stem.replace("_", ":", 1)
+            decoded = self._decode_storage_key(path.stem)
```
Comment (chengyongru) [inline nanobot/session/manager.py:868]: Old lossy filenames like `telegram_abc` can coincidentally be valid base64url. If `_decode_storage_key` returns a valid UTF-8 string, the `decoded or ...` fallback is skipped and a garbled key is used. This only affects sessions with missing/corrupt metadata, so impact is low, but a guard (e.g. checking that the decoded key contains `:`) would make it more robust.

**Review by chengyongru — COMMENTED** [review]
Automated PR review by nanobot. This is not a human maintainer review or approval.

Well-structured fix. The base64url encoding eliminates the `telegram:a_b` / `telegram:a:b` collision cleanly, and the migration chain (lossy path → legacy global path → new path) with `break` on first match is correct. Tests cover collision resistance, migration, and fallback thoroughly.

One minor edge case in `_decode_storage_key` during `list_sessions`: old lossy filenames (e.g. `telegram_abc`) may happen to be valid base64url and decode to arbitrary bytes. If those bytes are valid UTF-8, the decoded key would be wrong (binary gibberish instead of `telegram:abc`). The metadata-first lookup (`data.get("key")`) mitigates this for well-formed sessions, but sessions with missing/corrupt metadata would show a garbled key in the list.

**Review by Re-bin — APPROVED** [review]
LGTM now. The base64url storage key fixes the real session-file collision without changing the public `safe_key()` helper that WebUI/HTTP paths still depend on.

I also pushed a small follow-up to respect the stored metadata key before migrating an old lossy filename. That closes the upgrade edge where `telegram:a_b` could otherwise be moved into `telegram:a:b` if the colliding legacy file was loaded through the wrong key first. This keeps the migration compatible while avoiding fresh cross-session contamination.

Yongru's list-session fallback observation is acceptable after the metadata-first path, and the new stored-key guard covers the more dangerous migration case.

Validated locally:
- `python -m pytest tests/agent/test_session_collision.py tests/agent/test_session_delete.py::test_storage_key_matches_internal_path tests/webui/test_session_list_index.py::test_session_manager_list_sessions_fallback_time_when_missing -q`
- `python -m ruff check nanobot/session/manager.py tests/agent/test_session_collision.py`

GitHub Actions is green across the full matrix.
